subsys/fs/nvs/nvs.c

```diff
@@ -805,11 +805,12 @@ ssize_t nvs_write(struct nvs_fs *fs, u16_t id, const void *data, size_t len)
 			goto end;
 		}
 
-		sector_freespace = fs->ate_wra - fs->data_wra;
+		sector_freespace = (s32_t)fs->ate_wra - fs->data_wra;
 
 		/* Leave space for delete ate */
-		if (sector_freespace >= data_size + ate_size) {
-
+		if ((sector_freespace >= (s32_t)(data_size + ate_size)) ||
+		    /* accommodate the delete ate */
+		    (len == 0 && sector_freespace >= 0)) {
```

Comment:
I don't think what is done here is correct, when data is written 2 ate's should be available, one for a possible delete and one for an empty ate. This empty ate is needed whenever a write to the sector end would fail, in that case the empty ate becomes the only way to determine the last written ate.

Reply:
A simple fix:
```
if (len == 0) {
    sector_freespace += ate_size;
}
if (sector_freespace >= data_size + 2 * ate_size) {
```

Reply:
So there was a mistake with previous implementation (because it didn't take empty ATE into account) ?

```if (sector_freespace >= data_size + ate_size) {```

Reply:
@kapi-no, yes and no: this is a border case where all entries would fit exactly, in most cases there will be sufficient place to add the delete. It is also questionable if this functionality is needed, with the correction applied there always is the possibility (even in the border case) to delete entries and free up space in the fs, in the previous implementation the fs would be full and you could only erase it to free up space. I am unsure what is best, if the delete functionality is always available and used it might go trough the flash lifetime very fast.

Reply:
I can said that my current fix is proper. Previously case of writing entry with data was properly served as well. Problem touched only delete ATE.

Reply:
> I don't think what is done here is correct, when data is written 2 ate's should be available, one for a possible delete and one for an empty ate. This empty ate is needed whenever a write to the sector end would fail, in that case the empty ate becomes the only way to determine the last written ate.

This is new idea that one empty ATE should be keep always.
Maximum filling for the sector will then looks like:
```
|n* data|empty-ATE|delete-ATE|k * ATE|
```

Reply:
> This is new idea that one empty ATE should be keep always.

Or just keeping place for the last delete ATE was not asserted - then it might been original idea.

Reply:
I would update design documentation (i think we don't have a such) about that.

Reply:
@nvlsianpu, I don't quite get what your decision would be:
a. Always keep space for a delete: then you would need to check that there are 2 ate's available when a data write is performed (the first part of the if clause).

or

b. Don't keep space for a delete, then you can keep the old code.

Reply:
@Laczen For sure I always want to be able to delete something, even if nvs is full. What I done was with assumption that the one empty ate is unneeded (I think that always better to have this 8 byte). Honestly I was thinking that empty ate was reserved to be used for the last delete, and last delete was not possible by a bug. So I'm investigating the code whether I can really get rid of empty ATE requirement (even that I worked hard trying fix that). If I don't find anything better I will go your way.

Reply:
@nvlsianpu, after going through the code again I realized that I was wrong regarding the empty ate, this is only needed during the startup (to stop addr -= ate_size). Your method should work. I would apply a small change and use a required_space calculation instead of free space (line 808):
```
	u16_t required_space;
		...

		required_space = 0;
		if (data_size) {
			/* Leave space for delete ate */
			required_space += data_size + ate_size;
		}
		if (fs->ate_wra >= fs->data_wra + required_space) {

			rc = nvs_flash_wrt_entry(fs, id, data, len);
			if (rc) {
				goto end;
			}
			break;
		}
	...
```

Reply:
I Applied this suggestion before the while loop.